File: Data Structures and Algorithms/AI/Classification Metrics/AUC-ROC/Python/auc-roc.py
import numpy as np

def roc_curve(truth_labels, predicted_probs):
    """
    Computes the Receiver Operating Characteristic (ROC) curve.

    @param truth_labels: List or array of ground truth binary labels (0 or 1).
    @param predicted_probs: List or array of predicted probabilities (floats in [0,1]).

    @return: Two lists - true positive rates (TPRs) and false positive rates (FPRs)
             computed at different classification thresholds.
    """
    thresholds = [0.1 * i for i in range(11)]
    tprs, fprs = [], []

    for threshold in thresholds:
        tp = fp = tn = fn = 0
        predictions = [1 if p >= threshold else 0 for p in predicted_probs]

        for i in range(len(truth_labels)):
            if predictions[i] == 1 and truth_labels[i] == 1:
                tp += 1
            elif predictions[i] == 1 and truth_labels[i] == 0:
                fp += 1
            elif predictions[i] == 0 and truth_labels[i] == 0:
                tn += 1
            elif predictions[i] == 0 and truth_labels[i] == 1:
                fn += 1

        tpr = tp / (tp + fn) if (tp + fn) > 0 else 0.0
        fpr = fp / (fp + tn) if (fp + tn) > 0 else 0.0

        tprs.append(tpr)
        fprs.append(fpr)

    return tprs, fprs


def compute_aucroc(tprs, fprs):
    """
    Computes the Area Under the ROC Curve (AUC-ROC) using the trapezoidal rule.

    @param tprs: List of true positive rates from the ROC curve.
    @param fprs: List of false positive rates from the ROC curve.

    @return: A float representing the AUC-ROC value.
    """
    aucroc = 0
    for i in range(1, len(tprs)):
        aucroc += 0.5 * abs(fprs[i] - fprs[i - 1]) * (tprs[i] + tprs[i - 1])
    
    return aucroc

File: Data Structures and Algorithms/AI/Classification/Classification Metrics/AUC-ROC/Cpp/auc-roc.cpp
#include <vector>
#include <cmath>
#include <iomanip>
#include <tuple>
#include <type_traits>

/**
 * @brief Computes the ROC curve.
 *
 * @tparam T Numeric type for truth labels (e.g., int, double)
 * @tparam U Numeric type for predicted probabilities (e.g., float, double)
 * @param truth_labels Vector of ground truth binary labels.
 * @param predicted_probs Vector of predicted probabilities (0 to 1).
 * @return std::tuple<std::vector<double>, std::vector<double>> TPRs and FPRs.
 */
template <
    typename T, typename U,
    typename std::enable_if<std::is_arithmetic<T>::value && std::is_arithmetic<U>::value, int>::type = 0
>
std::tuple<std::vector<double>, std::vector<double>>
roc_curve(const std::vector<T>& truth_labels,
          const std::vector<U>& predicted_probs) {
    
    std::vector<double> thresholds;
    for (int i = 0; i <= 10; ++i)
        thresholds.push_back(i * 0.1);

    std::vector<double> tprs, fprs;

    for (double threshold : thresholds) {
        int tp = 0, fp = 0, tn = 0, fn = 0;

        for (std::size_t i = 0; i < truth_labels.size(); ++i) {
            int prediction = predicted_probs[i] >= threshold ? 1 : 0;
            if (prediction == 1 && truth_labels[i] == 1) {
                ++tp;
            }
            else if (prediction == 1 && truth_labels[i] == 0) {
                ++fp;
            }
            else if (prediction == 0 && truth_labels[i] == 0) {
                ++tn;
            }
            else if (prediction == 0 && truth_labels[i] == 1) {
                ++fn;
            }
        }

        double tpr = (tp + fn) > 0 ? static_cast<double>(tp) / (tp + fn) : 0.0;
        double fpr = (fp + tn) > 0 ? static_cast<double>(fp) / (fp + tn) : 0.0;

        tprs.push_back(tpr);
        fprs.push_back(fpr);
    }

    return std::make_tuple(tprs, fprs);
}

/**
 * @brief Computes the AUC-ROC from TPR and FPR vectors.
 *
 * @param tprs Vector of TPRs.
 * @param fprs Vector of FPRs.
 * @return double AUC-ROC value.
 */
double compute_aucroc(const std::vector<double>& tprs, const std::vector<double>& fprs) {
    double auc = 0.0;
    for (std::size_t i = 1; i < tprs.size(); ++i) {
        double width = std::abs(fprs[i] - fprs[i - 1]);
        double height = (tprs[i] + tprs[i - 1]) / 2.0;
        auc += width * height;
    }
    
    return auc;
}